### PR TITLE
Fix expm1 backward for large negative inputs

### LINKED ISSUE #192406: torch.expm1 backward returns zero gradient at -40.0 although the float64 derivative is finite
### 🐛 Describe the bug

`torch.autograd` returns a zero gradient for `torch.expm1` at a `float64` input where the mathematically correct derivative is finite and representable.

The derivative of `expm1(x)` is `exp(x)`. For `x = -40.0`, the true derivative is approximately `4.248354255291589e-18`, which is a normal finite `float64` value. However, PyTorch returns `0.0` for the gradient.

This issue reproduces on CPU, so CUDA availability does not affect the result.

```python
import torch

torch.set_default_dtype(torch.float64)

x = torch.tensor(-40.0, requires_grad=True)

y = torch.expm1(x)
y.backward()

actual = x.grad
expected = torch.tensor(4.248354255291589e-18, dtype=torch.float64)

print("actual:")
print(actual)
print("expected:")
print(expected)
```

### Actual vs expected result

```text
actual:
tensor(0.)

expected:
tensor(4.2484e-18)
```

The discrepancy occurs in the backward pass. `torch.expm1(-40.0)` itself is close to `-1`, but its derivative should be `exp(-40.0)`, not `0.0`.

### Versions

```text
PyTorch version: 2.12.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
Python version: 3.13.13 | packaged by Anaconda, Inc.
Python platform: Linux x86_64

Is CUDA available: False
GPU models:
  GPU 0: NVIDIA RTX 6000 Ada Generation
  GPU 1: NVIDIA RTX 6000 Ada Generation
Nvidia driver version: 570.211.01

CPU:
  Model name: AMD Ryzen Threadripper PRO 7985WX 64-Cores
  CPU(s): 128

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[pip3] triton==3.7.0
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by NishchalTendulkar**
## Summary

Fixes #192406.

The backward formula for `torch.expm1` currently computes the gradient as:

```python
grad * (result.conj() + 1)
```

Although mathematically equivalent to `exp(self)`, this formulation suffers from catastrophic cancellation for large negative inputs because `expm1(x)` approaches `-1`. As a result, `result + 1` can round to zero, producing an incorrect gradient.

This change computes the gradient directly as:

```python
grad * self.exp().conj()
```

which avoids the loss of precision while remaining mathematically equivalent.

A regression test has also been added to verify that the backward gradient of `torch.expm1(-40.0)` matches `torch.exp(-40.0)` for `float64`.

## Test Plan

```bash
python test/test_autograd.py -k expm1
```

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -125,6 +125,15 @@ def tearDown(self):
         torch.autograd._force_original_view_tracking(False)
         super().tearDown()
 
+    def test_expm1_backward_large_negative(self):
```
Comment (Skylion007) [inline test/test_autograd.py]: Should this be an opinfo input is my only complaint

Comment (NishchalTendulkar) [reply test/test_autograd.py]: Thanks for the review! Would you prefer moving this regression into the expm1 OpInfo instead of keeping it as a standalone autograd test? I'm happy to update the PR accordingly.

Comment (Skylion007) [reply test/test_autograd.py]: Yep

Comment (NishchalTendulkar) [reply test/test_autograd.py]: Thanks! I've moved the regression into the expm1 OpInfo by adding a dedicated sample_inputs_expm1 and removed the standalone autograd test.

**Review by soulitzer — COMMENTED** [review]
Thanks for the PR!

Yes this improves numerics, but it also regresses memory and perf:
- if expm1().sin() would now save an additional buffer
- expm1_() would trigger a clone to save the input
- computing exp is harder than add (although both are mem bound, so it might be noise)

It might be a worthy trade off though.

**Comment by NishchalTendulkar** [comment]
Thanks for the feedback! I investigated the CI failures after the previous update and found they were caused by the new scalar OpInfo sample being used by the sparse CSR OpInfo tests, which expect 2D inputs. I updated the regression sample to use a 2D tensor while keeping the same -40.0 regression value, and verified locally with:

python test/test_ops.py -k expm1
python test/test_sparse_csr.py -k expm1
python test/test_sparse_csr.py -k test_sparse_csr_unary_inplace_expm1_cpu_float64

**Comment by Skylion007** [comment]
@claude is there a way to handle better numerics here without the additional buffer?

**Comment by NishchalTendulkar** [comment]
Hi! Just checking in. The CI has passed, and I'm happy to make any additional changes if needed. Please let me know if there's anything else I should address. Thanks!

**Review by soulitzer — CHANGES_REQUESTED** [review]
I commented on the issue, it's not clear we should do this unless there's a good reason supporting this additional range matters.
